pymc3/sampling.py

```diff
@@ -1145,40 +1146,58 @@ def sample_posterior_predictive(trace,
                       DeprecationWarning)
     if vars is None:
         vars = model.observed_RVs
+    if var_names is None:
+        var_names = [rv.name for rv in vars]
 
     if random_seed is not None:
         np.random.seed(random_seed)
 
-    indices = np.arange(samples)
+    size = to_tuple(size)
 
-    
-    if progressbar:
-        indices = tqdm(indices, total=samples)
-
-    ppc_trace_t = _DefaultTrace(samples)
-    try:
-        for idx in indices:
-            if nchain > 1:
-                chain_idx, point_idx = np.divmod(idx, len_trace)
-                param = trace._straces[chain_idx % nchain].point(point_idx)
-            else:
-                param = trace[idx % len_trace]
-
-            values = draw_values(vars, point=param, size=size)
-            for k, v in zip(vars, values):
-                ppc_trace_t.insert(k.name, v, idx)
-
-    except KeyboardInterrupt:
-        pass
-
-    finally:
-        if progressbar:
-            indices.close()
-
-    ppc_trace = ppc_trace_t.trace_dict
-    if keep_size:
-        for k, ary in ppc_trace.items():
-            ppc_trace[k] = ary.reshape((nchain, len_trace, *ary.shape[1:]))
+    # Now we extract all the data from the trace and build a single dictionary,
```

Comment:
In a mod on `master` I revised this so we don't go through the intermediate step of building a Python list/array, and jump directly into populating the numpy array of samples.  Could we pull that over here?  I believe there are memory issues in building these temporary python data structures.